<a href="https://colab.research.google.com/github/NatalyBarbosa/Data-Science/blob/main/06_Pandas_manipulacao.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
dados = pd.read_json('dados_hospedagem.json')
dados.head()

info_moveis
0  {'avaliacao_geral': '10.0', 'experiencia_local...
1  {'avaliacao_geral': '10.0', 'experiencia_local...
2  {'avaliacao_geral': '10.0', 'experiencia_local...
3  {'avaliacao_geral': '10.0', 'experiencia_local...
4  {'avaliacao_geral': '10.0', 'experiencia_local...

In [ ]:
#normalizando a coluna "info-moveis"

dados = pd.json_normalize(dados['info_moveis'])
dados

avaliacao_geral experiencia_local max_hospedes  \
0             10.0                --            1   
1             10.0                --           10   
2             10.0                --           11   
3             10.0                --           12   
4             10.0                --           14   
..             ...               ...          ...   
65             nan                --            5   
66             nan                --            6   
67             nan                --            7   
68             nan                --            8   
69             nan                --            9   

                                      descricao_local  \
0   [This clean and comfortable one bedroom sits r...   
1   [Welcome to the heart of the 'Ballard Brewery ...   
2   [New modern house built in 2013.  Spectacular ...   
3   [Our NW style home is 3200+ sq ft with 3 level...   
4   [Perfect for groups. 2 bedrooms, full bathroom...   
..                                                ...   
65  [Our cozy little bungalow is the perfect place...   
66  [Bright clean 2 bedroom 1 bath appartment in t...   
67  [Beautiful end-unit townhome with lots of dayl...   
68  [Heart of Ballard new townhome with 4 BRs + ba...   
69  [This business-themed modern home features:  *...   

                                 descricao_vizinhanca  \
0   [Lower Queen Anne is near the Seattle Center (...   
1   [--, Capital Hill is the heart of Seattle, bor...   
2   [Upper Queen Anne is a charming neighborhood f...   
3   [The Views from our top floor!  Wallingford ha...   
4   [Safeway grocery store within walking distance...   
..                                                ...   
65  [The Queen Anne neighborhood is one of the mos...   
66  [Ballard is the most interesting and ecclectic...   
67  [Cherry blossoms (Sakura) in the spring, every...   
68  [--, --, The suite is conveniently located off...   
69  [Your hosts made Madison Valley their home whe...   

                                 quantidade_banheiros  \
0   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                   [2, 3, 2, 3, 3, 3, 2, 1, 2, 2, 2]   
2                                                 [4]   
3                            [3, 3, 3, 3, 3, 3, 3, 3]   
4                                              [2, 3]   
..                                                ...   
65  [1, 3, 2, 2, 2, 1, 2, 1, 1, 1, 2, 3, 2, 2, 2, ...   
66  [1, 1, 3, 2, 1, 3, 2, 2, 1, 2, 2, 2, 1, 2, 1, ...   
67                     [4, 3, 1, 2, 2, 1, 2, 1, 2, 2]   
68  [2, 2, 3, 1, 1, 2, 1, 3, 2, 1, 2, 1, 2, 2, 2, ...   
69                                             [2, 1]   

                                   quantidade_quartos  \
0   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                   [3, 4, 2, 3, 3, 3, 3, 3, 3, 4, 3]   
2                                                 [5]   
3                            [6, 6, 5, 5, 5, 5, 4, 4]   
4                                              [2, 6]   
..                                                ...   
65  [2, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, ...   
66  [2, 3, 3, 2, 2, 3, 2, 3, 2, 2, 2, 2, 4, 3, 3, ...   
67                     [3, 4, 3, 3, 3, 3, 3, 3, 3, 4]   
68  [4, 2, 3, 4, 3, 3, 4, 4, 3, 5, 2, 3, 3, 4, 5, ...   
69                                             [3, 4]   

                                     quantidade_camas  \
0   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                   [5, 6, 8, 3, 3, 5, 4, 5, 6, 7, 4]   
2                                                 [7]   
3                            [6, 6, 7, 8, 7, 7, 6, 6]   
4                                              [3, 9]   
..                                                ...   
65  [3, 2, 2, 3, 3, 2, 3, 2, 2, 3, 3, 3, 2, 3, 2, ...   
66  [3, 3, 2, 3, 3, 3, 2, 3, 3, 3, 2, 2, 4, 4, 3, ...   
67                     [3, 5, 1, 5, 3, 2, 3, 4, 4, 4]   
68  [4, 4, 3, 3, 5, 5, 5, 5, 3, 5, 4, 5, 5, 4, 1, ...   

## Coletando os nomes das colunas

Para desagrupar esses dados, podemos utilizar o método .explode(), que separa as informações em linhas. Ele requer que enviemos o nome das colunas que serão transformadas, então o primeiro passo é coletar esses nomes.

Para isso, criaremos uma lista chamada colunas onde transformaremos os valores em listas:

In [ ]:
colunas = list(dados.columns)
colunas

['avaliacao_geral',
 'experiencia_local',
 'max_hospedes',
 'descricao_local',
 'descricao_vizinhanca',
 'quantidade_banheiros',
 'quantidade_quartos',
 'quantidade_camas',
 'modelo_cama',
 'comodidades',
 'taxa_deposito',
 'taxa_limpeza',
 'preco']

## Usando o .explode() para desagrupar dados
**negrito**
Como queremos que nosso dataframe não possua dados agrupados em listas, vamos reatribuir essas informações para os dados já destrinchados em linhas. Para isso, passamos o método .explode() especificando as colunas de dados que queremos que sejam separados em linhas. Ou seja, a lista colunas a partir do índice 3, seguido de :, que indica que deve ir até o último valor da lista.

In [ ]:
dados = dados.explode(colunas[3:]) #conta a partir do índice 3
dados #desagrupados

avaliacao_geral experiencia_local max_hospedes  \
0             10.0                --            1   
0             10.0                --            1   
0             10.0                --            1   
0             10.0                --            1   
0             10.0                --            1   
..             ...               ...          ...   
68             nan                --            8   
68             nan                --            8   
68             nan                --            8   
69             nan                --            9   
69             nan                --            9   

                                      descricao_local  \
0   This clean and comfortable one bedroom sits ri...   
0   Our century old Upper Queen Anne house is loca...   
0   Cozy room in two-bedroom apartment along the l...   
0   Very lovely and cozy room for one. Convenientl...   
0   The “Studio at Mibbett Hollow' is in a Beautif...   
..                                                ...   
68  Beautiful craftsman home in the historic Wedgw...   
68  Located in a very easily accessible area of Se...   
68  This home is fully furnished and available wee...   
69  This business-themed modern home features:  *H...   
69  This welcoming home is in the quiet residentia...   

                                 descricao_vizinhanca quantidade_banheiros  \
0   Lower Queen Anne is near the Seattle Center (s...                    1   
0   Upper Queen Anne is a really pleasant, unique ...                    1   
0   The convenience of being in Seattle but on the...                    1   
0   Ballard is lovely, vibrant and one of the most...                    1   
0                                                  --                    1   
..                                                ...                  ...   
68                                                 --                    3   
68  Quiet, dead end street near I-5. The proximity...                    2   
68                                                 --                    1   
69  Your hosts made Madison Valley their home when...                    2   
69  Despite the quiet, leafy streets of this resid...                    1   

   quantidade_quartos quantidade_camas    modelo_cama  \
0                   1                1       Real Bed   
0                   1                1          Futon   
0                   1                1          Futon   
0                   1                1  Pull-out Sofa   
0                   1                1       Real Bed   
..                ...              ...            ...   
68                  4                5       Real Bed   
68                  4                4       Real Bed   
68                  3                4       Real Bed   
69                  3                6       Real Bed   
69                  4                6       Real Bed   

                                          comodidades taxa_deposito  \
0   {Internet,"Wireless Internet",Kitchen,"Free Pa...            $0   
0   {TV,Internet,"Wireless Internet",Kitchen,"Free...            $0   
0   {TV,Internet,"Wireless Internet",Kitchen,"Free...            $0   
0   {Internet,"Wireless Internet",Kitchen,"Free Pa...            $0   
0   {"Wireless Internet",Kitchen,"Free Parking on ...            $0   
..                                                ...           ...   
68  {TV,"Cable TV",Internet,"Wireless Internet","A...     $1,000.00   
68  {TV,"Cable TV",Internet,"Wireless Internet",Ki...            $0   
68  {TV,"Cable TV",Internet,"Wireless Internet","A...            $0   
69  {TV,"Cable TV",Internet,"Wireless Internet","A...     $1,000.00   
69  {TV,Internet,"Wireless Internet",Kitchen,"Free...       $500.00   

   taxa_limpeza    preco  
0            $0  $110.00  
0            $0   $45.00  
0            $0   $55.00  
0        $20.00   $52.00  
0        $15.00   $85.00  
..          ...      ...  
68      $178.00  

## Resetando os índices para ordenar os dados

Para isso, usaremos o método .reset_index() passando os parâmetros inplace=True e drop=True. O primeiro substitui os índices no dataframe, já o segundo trata-se da criação de uma coluna com os índices antigos, porém, sendo True, não teremos essa criação. Por fim, passamos a visualização dos primeiros registros com dados.head().

In [ ]:
dados.reset_index(inplace = True,drop=True) #reseta index antigos e coloca novos índices ordenados
dados.head()

avaliacao_geral experiencia_local max_hospedes  \
0            10.0                --            1   
1            10.0                --            1   
2            10.0                --            1   
3            10.0                --            1   
4            10.0                --            1   

                                     descricao_local  \
0  This clean and comfortable one bedroom sits ri...   
1  Our century old Upper Queen Anne house is loca...   
2  Cozy room in two-bedroom apartment along the l...   
3  Very lovely and cozy room for one. Convenientl...   
4  The “Studio at Mibbett Hollow' is in a Beautif...   

                                descricao_vizinhanca quantidade_banheiros  \
0  Lower Queen Anne is near the Seattle Center (s...                    1   
1  Upper Queen Anne is a really pleasant, unique ...                    1   
2  The convenience of being in Seattle but on the...                    1   
3  Ballard is lovely, vibrant and one of the most...                    1   
4                                                 --                    1   

  quantidade_quartos quantidade_camas    modelo_cama  \
0                  1                1       Real Bed   
1                  1                1          Futon   
2                  1                1          Futon   
3                  1                1  Pull-out Sofa   
4                  1                1       Real Bed   

                                         comodidades taxa_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...            $0   
1  {TV,Internet,"Wireless Internet",Kitchen,"Free...            $0   
2  {TV,Internet,"Wireless Internet",Kitchen,"Free...            $0   
3  {Internet,"Wireless Internet",Kitchen,"Free Pa...            $0   
4  {"Wireless Internet",Kitchen,"Free Parking on ...            $0   

  taxa_limpeza    preco  
0           $0  $110.00  
1           $0   $45.00  
2           $0   $55.00  
3       $20.00   $52.00  
4       $15.00   $85.00

## Verificando os tipos de dados

É possível notar as características dos dados do nosso dataframe: há textos e números inteiros e float. Mas precisamos garantir que estejam com a tipagem correta para que possamos utilizá-los em tratamentos futuros.

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   avaliacao_geral       3818 non-null   object
 1   experiencia_local     3818 non-null   object
 2   max_hospedes          3818 non-null   object
 3   descricao_local       3818 non-null   object
 4   descricao_vizinhanca  3818 non-null   object
 5   quantidade_banheiros  3818 non-null   object
 6   quantidade_quartos    3818 non-null   object
 7   quantidade_camas      3818 non-null   object
 8   modelo_cama           3818 non-null   object
 9   comodidades           3818 non-null   object
 10  taxa_deposito         3818 non-null   object
 11  taxa_limpeza          3818 non-null   object
 12  preco                 3818 non-null   object
dtypes: object(13)
memory usage: 387.9+ KB


Perceba que a coluna "Dtype" traz todos os dados como sendo do tipo object. Há três possibilidades para que um dado seja identificado assim:

  1.   por termos tipos diferentes de dados em uma mesma coluna;
  2.   quando os dados possuem um tipo específico que não foi identificado na leitura do Pandas;
  3.   todos os tipos são strings.

Para trabalharmos com estatística, por exemplo, precisamos que os dados numéricos sejam corretamente tipados como números. Sendo assim, quando temos várias colunas do tipo object, a **transformação da tipagem** precisa ser feita manualmente.

Começaremos a conversão dos nossos dados pelas colunas numéricas. Para isso, usaremos o método .astype(). Mas, primeiro, importaremos a biblioteca NumPy, que nos fornecerá um tipo de dado (inteiro e float) para podermos converter as colunas numéricas no nosso dataframe.

In [ ]:
import numpy as np

Tratando dados numéricos: convertendo para tipo inteiro

Iniciaremos o tratamento pela coluna "max_hospedes", que é de números inteiros.

O método .astype() nos permite converter colunas para um tipo específico de dado. Sendo assim, chamamos nosso dataframe dados e passamos a coluna que queremos converter entre colchetes ['max_hospedes']. Em seguida, chamamos o método .astype() passando o tipo de dado para o qual vamos alterar, no caso, np.int64, que é um tipo específico de inteiro com precisão de 64 bits.

In [ ]:
dados['max_hospedes']=dados['max_hospedes'].astype(np.int64)

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   avaliacao_geral       3818 non-null   object
 1   experiencia_local     3818 non-null   object
 2   max_hospedes          3818 non-null   int64 
 3   descricao_local       3818 non-null   object
 4   descricao_vizinhanca  3818 non-null   object
 5   quantidade_banheiros  3818 non-null   object
 6   quantidade_quartos    3818 non-null   object
 7   quantidade_camas      3818 non-null   object
 8   modelo_cama           3818 non-null   object
 9   comodidades           3818 non-null   object
 10  taxa_deposito         3818 non-null   object
 11  taxa_limpeza          3818 non-null   object
 12  preco                 3818 non-null   object
dtypes: int64(1), object(12)
memory usage: 387.9+ KB


Para isso, criaremos uma variável col_numericas, que será uma lista com as colunas cujos dados devem ser do tipo inteiro. São elas "quantidade_banheiros", "quantidade_quartos", "quantidade_camas".


In [ ]:
col_numericas = ['quantidade_banheiros','quantidade_quartos','quantidade_camas']

In [ ]:
dados[col_numericas] = dados[col_numericas].astype(np.int64) #Note que as colunas transformadas estão definidas como sendo do tipo int64:

In [ ]:
dados['avaliacao_geral'] = dados['avaliacao_geral'].astype(np.float64)

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   avaliacao_geral       3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descricao_local       3818 non-null   object 
 4   descricao_vizinhanca  3818 non-null   object 
 5   quantidade_banheiros  3818 non-null   int64  
 6   quantidade_quartos    3818 non-null   int64  
 7   quantidade_camas      3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  taxa_deposito         3818 non-null   object 
 11  taxa_limpeza          3818 non-null   object 
 12  preco                 3818 non-null   object 
dtypes: float64(1), int64(4), object(8)
memory usage: 387.9+ KB


Cada elemento dessa coluna é uma string, então podemos fazer uma função que passe linha por linha para remover os caracteres.

Para isso, usaremos o método .apply() com lambda, que aplica uma função específica a cada elemento, no caso, a remoção dos caracteres.

Dentro de .apply(), passamos uma função lambda que lerá cada elemento x, ou seja, os dados de cada linha. Feito isso, usaremos o método .replace() para substituir os caracteres de cifrão e vírgula por nada, ou seja, por uma string vazia. Depois, usaremos .strip() para remover qualquer caractere antes ou depois do elemento.

In [ ]:
dados['preco'] = dados['preco'].apply(lambda x: x.replace('$','').replace(',','').strip())

O applymap()método permite aplicar uma ou mais funções ao objeto DataFrame.

Vamos transformar os valores dessas colunas para float utilizando o método .astype():



In [ ]:
dados['preco'] = dados['preco'].astype(np.float64)

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   avaliacao_geral       3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descricao_local       3818 non-null   object 
 4   descricao_vizinhanca  3818 non-null   object 
 5   quantidade_banheiros  3818 non-null   int64  
 6   quantidade_quartos    3818 non-null   int64  
 7   quantidade_camas      3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  taxa_deposito         3818 non-null   object 
 11  taxa_limpeza          3818 non-null   object 
 12  preco                 3818 non-null   float64
dtypes: float64(2), int64(4), object(7)
memory usage: 387.9+ KB


Note que, assim como a coluna "preco", temos dados com cifrão e vírgula. Porém, para convertê-los, não usaremos o método .apply(), pois ele é utilizado para linhas e colunas.

## Aplicar o .applymap()

Quando tratamos da conversão de um dataframe, no qual passaremos por cada elemento, é mais viável utilizarmos o método .applymap().

Dentro dos parênteses deste método, aplicamos o mesmo processo aplicado à "preco". Ou seja, a função lambda e os métodos .replace() e .strip():

In [ ]:
dados[['taxa_deposito','taxa_limpeza']] = dados[['taxa_deposito','taxa_limpeza']].applymap(lambda x: x.replace('$','').replace(',','').strip())

In [ ]:
dados[['taxa_deposito','taxa_limpeza']] = dados[['taxa_deposito','taxa_limpeza']].astype(np.float64)

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   avaliacao_geral       3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descricao_local       3818 non-null   object 
 4   descricao_vizinhanca  3818 non-null   object 
 5   quantidade_banheiros  3818 non-null   int64  
 6   quantidade_quartos    3818 non-null   int64  
 7   quantidade_camas      3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  taxa_deposito         3818 non-null   float64
 11  taxa_limpeza          3818 non-null   float64
 12  preco                 3818 non-null   float64
dtypes: float64(4), int64(4), object(5)
memory usage: 387.9+ KB


Podemos tratar esses textos buscando criar uma estrutura mais acessível para realizar essa análise textual. Nessa aula, realizaremos a manipulação dos textos de cada coluna textual para criar uma estrutura de tokenização simples.

A **tokenização** é a divisão do texto em tokens, ou seja, unidades menores. Esses tokens podem ser palavras, sílabas, conjuntos de palavras e assim por diante, dependendo do tipo de aplicação.

Como faremos uma tokenização simples, separaremos o texto apenas por palavras. Ou seja, os tokens serão cada palavra do nosso texto.

## Manipulação dos textos

Podemos começar a estruturar a tokenização padronizando o nosso texto para letras minúsculas. Então, de volta ao Google Colab, o primeiro passo será manipular os dados da coluna descricao_local. Essa coluna será o foco inicial da nossa aula.

In [ ]:
dados['descricao_local'] = dados['descricao_local'].str.lower()

Podemos observar que temos letras maiúsculas nos textos. Para converter todos esses elementos para letras minúsculas, podemos utilizar o método str(). Com ele, conseguimos ler a nossa série como se ela fosse elementos de strings, de modo similar aos métodos apply e applymap.

Então, ao lado de dados['descricao_local'], colocamos um .str(). Isso já fará o processo de leitura dos elementos dessa coluna como strings. A partir disso, podemos aplicar os métodos originários do tipo string nesse dataframe. Então, se queremos transformar essas letras para minúsculas, utilizaremos o método lower():

In [ ]:
dados.head()

avaliacao_geral experiencia_local  max_hospedes  \
0             10.0                --             1   
1             10.0                --             1   
2             10.0                --             1   
3             10.0                --             1   
4             10.0                --             1   

                                     descricao_local  \
0  this clean and comfortable one bedroom sits ri...   
1  our century old upper queen anne house is loca...   
2  cozy room in two-bedroom apartment along the l...   
3  very lovely and cozy room for one. convenientl...   
4  the “studio at mibbett hollow' is in a beautif...   

                                descricao_vizinhanca  quantidade_banheiros  \
0  Lower Queen Anne is near the Seattle Center (s...                     1   
1  Upper Queen Anne is a really pleasant, unique ...                     1   
2  The convenience of being in Seattle but on the...                     1   
3  Ballard is lovely, vibrant and one of the most...                     1   
4                                                 --                     1   

   quantidade_quartos  quantidade_camas    modelo_cama  \
0                   1                 1       Real Bed   
1                   1                 1          Futon   
2                   1                 1          Futon   
3                   1                 1  Pull-out Sofa   
4                   1                 1       Real Bed   

                                         comodidades  taxa_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...            0.0   
1  {TV,Internet,"Wireless Internet",Kitchen,"Free...            0.0   
2  {TV,Internet,"Wireless Internet",Kitchen,"Free...            0.0   
3  {Internet,"Wireless Internet",Kitchen,"Free Pa...            0.0   
4  {"Wireless Internet",Kitchen,"Free Parking on ...            0.0   

   taxa_limpeza  preco  
0           0.0  110.0  
1           0.0   45.0  
2           0.0   55.0  
3          20.0   52.0  
4          15.0   85.0

In [ ]:
dados['descricao_local'][3169]

"built, run and supported by seattle tech and start up veterans, grokhome's focus is to create a supportive environment for smart people working on interesting projects, start ups and more. this listing is an upper bunk, in a 2-person shared room. *note: this fall, there will be major renovations happening on one kitchen and bathroom at a time. there will always be two other working kitchens and two working bathrooms in the house. we'll work to minimize the impact these renovations have on your stay. **this listing is only available to those working in the tech/science space. live in a hacker house, and immerse yourself in the seattle tech scene. you can expect to be surrounded by smart people solving big problems or working on something fun. we have frequent demo nights, and love when our guests share something they are passionate about. if you're new to the city, our deep ties to the seattle tech scene can help you get involved. expand your network, develop your ideas, and learn some

## Regex: removendo caracteres especiais

Dentro dos parênteses do método replace(), colocaremos aspas simples e passaremos como primeiro parâmetro o Regex responsável por encontrar todos os caracteres do texto - com exceção das letras, números, apóstrofo e hífen. O Regex é o seguinte:

[^a-zA-Z0-9\-\']

Vamos entender sua estrutura. Nós abrimos e fechamos colchetes e, dentro deles, começamos a expressão com um sinal circunflexo (^), que nega os próximos valores. Ou seja, colocamos em seguida os valores que ele não vai selecionar:

* a-z: todas as letras, do "a" ao "z", minúsculas;
* A-Z: todas as letras, do "A" ao "Z", maiúsculas;
* 0-9: todos os números, de 0 a 9;
* \-: hífen;
* \': apóstrofo.

Toda essa expressão diz: "selecione todos os caracteres, exceto as letras (maiúsculas ou minúsculas), números, hífen e apóstrofo".

Vamos adicionar uma vírgula depois dessa expressão e adicionar mais um par de aspas simples com um espaço entre elas. Assim, todos os caracteres especiais serão substituídos por um espaço.

Por fim, para conseguir utilizar esse código Regex, precisamos inserir o parâmetro de Regex dentro de replace(). Então adicionamos mais uma vírgula e escrevemos regex=True.

In [ ]:
dados['descricao_local'] = dados['descricao_local'].str.replace('[^a-zA-Z0-9\-\']',' ',regex=True)

Removendo hifens fora de palavras compostas
O nosso próximo passo será remover os hifens que não estejam conectando palavras compostas, ou seja, os hifens soltos que não tenham nenhum caractere antes ou depois deles. Há um Regex específico para realizar essa tarefa!

Primeiro, vamos copiar todo o texto a partir do sinal de igual da nossa atribuição e colar em uma nova célula (dados['descricao_local'].str.replace('(?<!\w)-(?!\w)', '', regex=True)).

No primeiro parâmetro de replace(), selecionamos todo o Regex que criamos e o apagamos para inserir um novo Regex, capaz de verificar se há elementos antes ou depois de determinado hífen e mantê-lo apenas se esse for o caso:

(?<!\w)-(?!\w)

Esse Regex tem uma estrutura diferente: entre parênteses, colocamos um sinal de interrogação (?), um sinal de menor (<), uma exclamação (!), barra () e a letra "w" minúscula. Depois dos primeiros parênteses, inserimos um hífen e abrimos novos parênteses, contendo os mesmos caracteres, exceto pelo sinal de menor. Ou seja:

(?<!\w): uma "lookbehind" negativa que verifica se há um caractere de palavra antes do hífen ou se há apenas um espaço em branco, resultando em "verdadeiro" se não houver nada;
(?!\w): uma lookahead negativa que verifica se há um caractere de palavra depois do hífen.
A combinação de lookbehind negativo e lookahead negativo garante que o hífen esteja entre espaços em branco ou no início ou fim de uma palavra. Nesse caso, esse hífen será selecionado por esse Regex e substituído por um espaço em branco.

In [ ]:
dados['descricao_local'] = dados['descricao_local'].str.replace('(?<!\w)-(?!\w)',' ',regex=True)

## Método split()

Para criar tokens, basta separar cada uma das palavras como elementos de uma lista. Para transformar uma string numa lista, usamos o método split(). Podemos aplicá-lo na nossa coluna por meio do str, que já nos acostumamos a utilizar!

Numa nova célula do Colab, vamos chamar a nossa coluna com dados['descricao_local']. Em seguida, inserimos .str para tratar elementos como strings. Depois disso, adicionamos o .split() - método capaz de dividir elementos textuais em uma lista, usando o espaço em branco como elemento separador.

In [ ]:
dados['descricao_local'] = dados['descricao_local'].str.split()
dados.head()

avaliacao_geral experiencia_local  max_hospedes  \
0             10.0                --             1   
1             10.0                --             1   
2             10.0                --             1   
3             10.0                --             1   
4             10.0                --             1   

                                     descricao_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
1  [our, century, old, upper, queen, anne, house,...   
2  [cozy, room, in, two-bedroom, apartment, along...   
3  [very, lovely, and, cozy, room, for, one, conv...   
4  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descricao_vizinhanca  quantidade_banheiros  \
0  Lower Queen Anne is near the Seattle Center (s...                     1   
1  Upper Queen Anne is a really pleasant, unique ...                     1   
2  The convenience of being in Seattle but on the...                     1   
3  Ballard is lovely, vibrant and one of the most...                     1   
4                                                 --                     1   

   quantidade_quartos  quantidade_camas    modelo_cama  \
0                   1                 1       Real Bed   
1                   1                 1          Futon   
2                   1                 1          Futon   
3                   1                 1  Pull-out Sofa   
4                   1                 1       Real Bed   

                                         comodidades  taxa_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...            0.0   
1  {TV,Internet,"Wireless Internet",Kitchen,"Free...            0.0   
2  {TV,Internet,"Wireless Internet",Kitchen,"Free...            0.0   
3  {Internet,"Wireless Internet",Kitchen,"Free Pa...            0.0   
4  {"Wireless Internet",Kitchen,"Free Parking on ...            0.0   

   taxa_limpeza  preco  
0           0.0  110.0  
1           0.0   45.0  
2           0.0   55.0  
3          20.0   52.0  
4          15.0   85.0

### **Tratamento dos dados**

Primeiramente, **removeremos as chaves e aspas** - caracteres que não nos ajudarão a tokenizar nossa coluna de comodidades. Já sabemos fazer isso usando o método `replace()` e o **Regex**: vamos identificar os caracteres em questão e substituí-los.

Então, como primeiro parâmetro de `replace()`, passamos o Regex `\{|}|\"` para encontrar:

- `{|` chave aberta;
- `}|` chave fechada;
- `"` aspas.
    - as barras verticais (`|`) representam uma escolha entre dois padrões diferentes.

Neste caso, podemos manter as vírgulas para utilizá-las como separador para dividir as palavras em uma lista na nossa tokenização.

Além disso, o que vai substituir esses caracteres é uma string vazia. Portanto, como segundo parâmetro, adicionamos apenas um par de aspas simples sem nada entre elas.

In [ ]:
dados['comodidades'] = dados['comodidades'].str.replace('\{|}|\"','',regex=True)

## Tokenização

Agora, criaremos a nossa lista de tokens. Faremos o mesmo que fizemos com a coluna descricao_local, utilizando o método split().

Para isso, vamos copiar a célula anterior até o str e colar em uma nova célula, adicionando .split() ao final. Dentro dos parênteses, vamos passar como parâmetro o nosso caractere separador, as vírgulas, entre aspas simples ','. Na linha abaixo, escrevemos dados.head() para verificar o resultado:

In [ ]:
dados['comodidades'] = dados['comodidades'].str.split(',')
dados.head()

avaliacao_geral experiencia_local  max_hospedes  \
0             10.0                --             1   
1             10.0                --             1   
2             10.0                --             1   
3             10.0                --             1   
4             10.0                --             1   

                                     descricao_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
1  [our, century, old, upper, queen, anne, house,...   
2  [cozy, room, in, two-bedroom, apartment, along...   
3  [very, lovely, and, cozy, room, for, one, conv...   
4  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descricao_vizinhanca  quantidade_banheiros  \
0  Lower Queen Anne is near the Seattle Center (s...                     1   
1  Upper Queen Anne is a really pleasant, unique ...                     1   
2  The convenience of being in Seattle but on the...                     1   
3  Ballard is lovely, vibrant and one of the most...                     1   
4                                                 --                     1   

   quantidade_quartos  quantidade_camas    modelo_cama  \
0                   1                 1       Real Bed   
1                   1                 1          Futon   
2                   1                 1          Futon   
3                   1                 1  Pull-out Sofa   
4                   1                 1       Real Bed   

                                         comodidades  taxa_deposito  \
0  [Internet, Wireless Internet, Kitchen, Free Pa...            0.0   
1  [TV, Internet, Wireless Internet, Kitchen, Fre...            0.0   
2  [TV, Internet, Wireless Internet, Kitchen, Fre...            0.0   
3  [Internet, Wireless Internet, Kitchen, Free Pa...            0.0   
4  [Wireless Internet, Kitchen, Free Parking on P...            0.0   

   taxa_limpeza  preco  
0           0.0  110.0  
1           0.0   45.0  
2           0.0   55.0  
3          20.0   52.0  
4          15.0   85.0

In [ ]:
dt_data = pd.read_json('/content/moveis_disponiveis.json')
dt_data.head()

id        data  vaga_disponivel preco
0  857  2016-01-04            False  None
1  857  2016-01-05            False  None
2  857  2016-01-06            False  None
3  857  2016-01-07            False  None
4  857  2016-01-08            False  None

In [ ]:
dt_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   id               365000 non-null  int64 
 1   data             365000 non-null  object
 2   vaga_disponivel  365000 non-null  bool  
 3   preco            270547 non-null  object
dtypes: bool(1), int64(1), object(2)
memory usage: 11.5+ MB


Temos quatro colunas: id, data, vaga_disponivel e preco. Dentre elas, aquela que mais nos interessa é a data.

Vemos que ela possui o formato "ano-mês-dia" (AAAA-MM-DD). Este é um formato padrão da estrutura do datetime, portanto, não precisamos nos preocupar com a formatação.

Nos resta verificar se os dados estão no tipo datetime. Para isso, selecionaremos a próxima célula e adicionaremos o método info().

dt_data.info()
COPIAR CÓDIGO
Pressionaremos "Shift+Enter" e veremos o retorno abaixo:

#	Column	Non-Null Count	Dtype
---	------	--------------	-----
0	id	365000 non-null	int64
1	data	365000 non-null	object
2	vaga_disponivel	365000 non-null	bool
3	preco	270547 non-null	object
Nele, notaremos que o tipo (Dtype) da segunda linha (que representa a coluna data) é object. Transformaremos esta coluna no tipo desejado — o temporal, ou seja, o datetime.

O Pandas fornece o método to_datetime() que nos permite transformar a coluna de um dataframe em datetime. Vamos chamar o Pandas pelo comando pd e utilizar este método em nossa célula.

In [ ]:
dt_data['data'] = pd.to_datetime(dt_data['data'])

In [ ]:
dt_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   id               365000 non-null  int64         
 1   data             365000 non-null  datetime64[ns]
 2   vaga_disponivel  365000 non-null  bool          
 3   preco            270547 non-null  object        
dtypes: bool(1), datetime64[ns](1), int64(1), object(1)
memory usage: 11.5+ MB


In [ ]:
dt_data.head()

id       data  vaga_disponivel preco
0  857 2016-01-04            False  None
1  857 2016-01-05            False  None
2  857 2016-01-06            False  None
3  857 2016-01-07            False  None
4  857 2016-01-08            False  None

Nas atividades desta aula, descobrimos o método chamado strftime, capaz de formatar os valores de data para o formato que quisermos.

In [ ]:
dt_data['data'].dt.strftime('%Y-%m')

0         2016-01
1         2016-01
2         2016-01
3         2016-01
4         2016-01
           ...   
364995    2016-12
364996    2016-12
364997    2016-12
364998    2017-01
364999    2017-01
Name: data, Length: 365000, dtype: object

In [ ]:
subset = dt_data.groupby(dt_data['data'].dt.strftime('%Y-%m'))['vaga_disponivel'].sum()
subset

data
2016-01    16543
2016-02    20128
2016-03    23357
2016-04    22597
2016-05    23842
2016-06    23651
2016-07    22329
2016-08    22529
2016-09    22471
2016-10    23765
2016-11    23352
2016-12    24409
2017-01     1574
Name: vaga_disponivel, dtype: int64